server/swap/swap.go

```diff
@@ -1132,145 +1158,172 @@ func (s *Swapper) makerRedeemStatus(mStatus *swapStatus, tAsset uint32) (makerRe
+	s.matchMtx.Lock()
+	defer s.matchMtx.Unlock()
 
-			// Record the end of this match's processing.
-			s.storage.SetMatchInactive(db.MatchID(match.Match))
+	for _, match := range s.matches {
+		checkMatch(match)
+	}
 
-			// If the at-fault order is a limit order, signal that if it is
-			// still on the book is should be unbooked, changed to revoked
-			// status, counted against the user's cancellation ratio, and a
-			// server-generated cancel order recorded.
-			if lo, isLimit := orderAtFault.(*order.LimitOrder); isLimit {
-				if s.unbookHook(lo) {
-					s.orders.canceled(orderAtFault) // set as off-book and failed
-				}
-			}
+	for _, match := range deletions {
+		delete(s.matches, match.ID())
+	}
+}
 
-			// That's one less active swap for this order, and a failure. If
-			// this order has no other active swaps, it will be removed from the
-			// order swap tracker by decrementActiveSwapCount since it is
-			// off-book and no new swap negotiations can begin for this order.
-			s.orders.swapFailure(orderAtFault)
-
-			// The other order now has one less active swap too.
-			if s.orders.swapSuccess(otherOrder) {
-				// TODO: We should count this as a successful swap, but should
-				// it only be a completed order with the extra stipulation that
-				// it had already completed another swap?
-				compTime := time.Now().UTC()
-				s.authMgr.RecordCompletedOrder(otherOrder.User(), otherOrder.ID(), compTime)
-				if err := s.storage.SetOrderCompleteTime(otherOrder, encode.UnixMilli(compTime)); err != nil {
-					if db.IsErrGeneralFailure(err) {
-						log.Errorf("fatal error with SetOrderCompleteTime for order %v: %v", otherOrder.UID(), err)
-					} else {
-						log.Warnf("SetOrderCompleteTime for %v: %v", otherOrder.UID(), err)
-					}
-				}
-			}
+// checkInaction scans the swapStatus structures relevant to the specified
+// asset. If a client is found to have not acted when required, a match may be
+// revoked and a penalty assigned to the user.
+func (s *Swapper) checkInactionBlockBased(assetID uint32) {
+	// If the DB is failing, do not penalize or attempt to start revocations.
+	if err := s.storage.LastErr(); err != nil {
+		log.Errorf("DB in failing state.")
+		return
+	}
 
-			// Penalize for failure to act.
-			//
-			// TODO: Arguably, this obviates the RecordCancel above since this
-			// closes the account before the possibility of a cancellation ratio
-			// penalty. I'm keeping it this way for now however since penalties
-			// may become less severe than account closure (e.g. temporary
-			// suspension, cool down, or order throttling), and restored
-			// accounts will still require a record of the revoked order.
-			s.authMgr.Penalize(orderAtFault.User(), account.FailureToAct)
+	var deletions []*matchTracker
+	// Do time.Since(event) with the same now time for each match.
+	now := time.Now()
+	tooOld := func(evt time.Time) bool {
+		return !evt.IsZero() && now.Sub(evt) >= s.bTimeout
+	}
```

Comment:
Why does this one need `!evt.IsZero()`.

Reply:
a zero Time means it hasn't been set yet, and thus not too old.